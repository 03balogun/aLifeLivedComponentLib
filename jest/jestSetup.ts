/* eslint-disable no-undef */
jest.useFakeTimers('legacy'); // Stops reference error from appearing
