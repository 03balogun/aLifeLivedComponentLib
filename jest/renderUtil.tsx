import React from 'react';
import { ThemeProvider } from '../src/themes/ThemeProvider';
import { render as rtlRender } from '@testing-library/react-native';

function render(ui, { ...renderOptions } = {}) {
  function Wrapper({ children }) {
    return <ThemeProvider>{children}</ThemeProvider>;
  }

  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
}

// re-export everything for convenience
export * from '@testing-library/react-native';

// override render method
export { render };
