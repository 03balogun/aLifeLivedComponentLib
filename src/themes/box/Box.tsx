import React from 'react';
import {
  createBox,
  createVariant,
  createRestyleComponent,
  VariantProps,
  BoxProps,
  RestyleFunctionContainer,
} from '@shopify/restyle';
import { ViewProps } from 'react-native';
import { Theme } from '../theme';

const variant = createVariant<Theme>({
  property: 'boxShadow',
  themeKey: 'shadows',
}) as RestyleFunctionContainer<
  VariantProps<Theme, 'shadows', 'boxShadow'>,
  Theme,
  'boxShadow',
  'shadows'
>;

export type StyledBoxProps = BoxProps<Theme> &
  ViewProps &
  VariantProps<Theme, 'shadows', 'boxShadow'> & {
    children?: React.ReactNode;
  };

export const Box = createRestyleComponent<StyledBoxProps, Theme>(
  [variant],
  createBox<Theme>()
);
