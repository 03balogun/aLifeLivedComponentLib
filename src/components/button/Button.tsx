import React, { ReactChild } from 'react';
import { Box } from '../../themes/box/Box';
import { useTheme } from '../../themes/hooks/useTheme';
import {
  StyledOpacityProps,
  SystemButton,
} from '../../themes/systemButton/SystemButton';
import { SystemSpinner } from '../../themes/systemSpinner/SystemSpinner';
import { Text } from '../../themes/text/Text';

interface IButtonProps {
  title: string;
  onPress: () => void;
  isDisabled?: boolean;
  variant?: 'cta';
  color?: string;
  isLoading?: boolean;
  endAdornmentComponent?: ReactChild;
}

export const Button: React.FC<IButtonProps & StyledOpacityProps> = ({
  title,
  onPress,
  isDisabled,
  color,
  isLoading,
  endAdornmentComponent,
  ...rest
}) => {
  return (
    <SystemButton
      backgroundColor={isDisabled ? 'backgroundContrastMid' : 'primary'}
      onPress={() => onPress()}
      padding="m"
      borderRadius="medium"
      width="100%"
      justifyContent="center"
      alignContent="center"
      disabled={isDisabled || isLoading}
      {...rest}
    >
      {isLoading ? (
        <SystemSpinner
          size="small"
          color="white"
          testID="QBYQCQ-button-spinner"
          animating
          hidesWhenStopped={true}
        />
      ) : (
        <Box
          flexDirection="row"
          alignContent="center"
          alignItems="center"
          justifyContent="center"
        >
          {endAdornmentComponent ? (
            <ButtonIconAdornment>{endAdornmentComponent}</ButtonIconAdornment>
          ) : null}
          <Text variant="intro" textAlign="center" color={color}>
            {title.toUpperCase()}
          </Text>
        </Box>
      )}
    </SystemButton>
  );
};

const ButtonIconAdornment: React.FC = React.memo(({ children }) => {
  const { spacing } = useTheme();
  return (
    <Box
      position="absolute"
      left={spacing.m}
      height="100%"
      alignItems="center"
      justifyContent="center"
      alignContent="center"
    >
      {children}
    </Box>
  );
});
