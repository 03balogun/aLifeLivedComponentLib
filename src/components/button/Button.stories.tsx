import React from 'react';

import { storiesOf } from '@storybook/react-native';

// Component
import { Button } from './Button';
import { action } from '@storybook/addon-actions';
import { CenterView } from '../centerView/CenterView';
import { Box } from '../../themes/box/Box';
import { Icon } from '../icon/Icon';

storiesOf('Button', module)
  .addDecorator(Story => <CenterView>{Story()}</CenterView>)
  .add('Default button', () => (
    <Box alignItems="center" width="100%">
      <Box width="90%" margin="m">
        <Button
          title="Test button that has a long long title"
          onPress={action('Button on press fired')}
          isLoading={false}
          isDisabled={false}
          color="backgroundContrastXLow"
          backgroundColor="secondary"
        />
      </Box>
      <Box width="90%" margin="m">
        <Button
          title="Loading button"
          onPress={action('Button on press')}
          isLoading={true}
          isDisabled={false}
        />
      </Box>
      <Box width="90%" margin="m">
        <Button
          title="Disabled button"
          onPress={action('Button on press')}
          isLoading={false}
          isDisabled={true}
        />
      </Box>
      <Box width="90%" margin="m">
        <Button
          title="Button with adornment"
          onPress={action('Button on press')}
          isLoading={false}
          isDisabled={true}
          endAdornmentComponent={<Icon icon="play" />}
        />
      </Box>
      <Box margin="m" width="50%">
        <Button
          title="Play all"
          onPress={action('Button on press')}
          isLoading={false}
          endAdornmentComponent={<Icon color="white" icon="play" />}
          color="white"
        />
      </Box>
    </Box>
  ));
