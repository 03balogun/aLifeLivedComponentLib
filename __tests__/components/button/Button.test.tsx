import React, { ReactChild } from 'react';
import { render, fireEvent } from './../../../jest/renderUtil';

// Component
import { Button } from '../../../src/components/button/Button';
import { Box } from '../../../src/themes/box/Box';

const mockOnPress = jest.fn();

const setup = ({
  isDisabled = false,
  isLoading = false,
  endAdornment = undefined,
}: {
  isLoading?: boolean;
  isDisabled?: boolean;
  endAdornment?: ReactChild;
}) =>
  render(
    <Button
      title="Test button"
      onPress={mockOnPress}
      isDisabled={isDisabled}
      isLoading={isLoading}
      endAdornmentComponent={endAdornment}
    />
  );

describe('components >> button ', () => {
  describe('Enabled button', () => {
    test('Should render basic button correctly', () => {
      const { toJSON } = setup({ isDisabled: false, isLoading: false });
      expect(toJSON()).toMatchSnapshot();
    });

    test('Should render title', () => {
      const { getByText } = setup({ isDisabled: false, isLoading: false });
      expect(getByText('TEST BUTTON')).not.toBeNull();
    });

    test('Should fire on press', () => {
      const { getByText } = setup({ isDisabled: false, isLoading: false });
      fireEvent.press(getByText('TEST BUTTON'));
      expect(mockOnPress).toBeCalledTimes(1);
    });
  });

  describe('Disabled button', () => {
    beforeAll(() => mockOnPress.mockClear());
    test('Should render basic button correctly', () => {
      const { toJSON } = setup({ isDisabled: true, isLoading: false });
      expect(toJSON()).toMatchSnapshot();
    });

    test('Should find title', () => {
      const { getByText } = setup({ isDisabled: true, isLoading: false });
      expect(getByText('TEST BUTTON')).toBeDefined();
    });

    test('Should not be able to fire on press action', () => {
      const { getByText } = setup({ isDisabled: true, isLoading: false });
      fireEvent.press(getByText('TEST BUTTON'));
      expect(mockOnPress).toBeCalledTimes(0);
    });
  });

  describe('Loading button', () => {
    beforeAll(() => mockOnPress.mockClear());
    test('Should render button in loading state', () => {
      const { toJSON } = setup({ isDisabled: false, isLoading: true });
      expect(toJSON()).toMatchSnapshot();
    });

    test('Should NOT render title', () => {
      const { queryByText } = setup({ isDisabled: false, isLoading: true });
      expect(queryByText('TEST BUTTON')).toBeNull();
    });

    test('Should NOT fire on press', () => {
      const { getByTestId } = setup({ isDisabled: false, isLoading: true });
      fireEvent.press(getByTestId('QBYQCQ-button-spinner'));
      expect(mockOnPress).toBeCalledTimes(0);
    });
  });

  describe('Given there is an endAdornment', () => {
    test('Should render child on left hand side of button', () => {
      const { getByTestId } = setup({
        endAdornment: <Box testID="end-adornment-test-id" />,
      });

      expect(getByTestId('end-adornment-test-id')).toBeDefined();
    });
  });
});
