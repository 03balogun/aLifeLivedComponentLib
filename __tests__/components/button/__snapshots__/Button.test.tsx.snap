// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components >> button  Disabled button Should render basic button correctly 1`] = `
<View
  accessibilityState={
    Object {
      "disabled": true,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignContent": "center",
      "backgroundColor": "#CBCBCB",
      "borderRadius": 8,
      "justifyContent": "center",
      "opacity": 1,
      "padding": 16,
      "width": "100%",
    }
  }
>
  <View
    style={
      Array [
        Object {
          "alignContent": "center",
          "alignItems": "center",
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "center",
        },
        Array [
          Object {},
        ],
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#282828",
            "display": "flex",
            "fontFamily": "SFUIText-Semibold",
            "fontSize": 16,
            "textAlign": "center",
          },
        ]
      }
    >
      TEST BUTTON
    </Text>
  </View>
</View>
`;

exports[`components >> button  Enabled button Should render basic button correctly 1`] = `
<View
  accessibilityState={
    Object {
      "disabled": false,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignContent": "center",
      "backgroundColor": "#67B6A8",
      "borderRadius": 8,
      "justifyContent": "center",
      "opacity": 1,
      "padding": 16,
      "width": "100%",
    }
  }
>
  <View
    style={
      Array [
        Object {
          "alignContent": "center",
          "alignItems": "center",
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "center",
        },
        Array [
          Object {},
        ],
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#282828",
            "display": "flex",
            "fontFamily": "SFUIText-Semibold",
            "fontSize": 16,
            "textAlign": "center",
          },
        ]
      }
    >
      TEST BUTTON
    </Text>
  </View>
</View>
`;

exports[`components >> button  Loading button Should render button in loading state 1`] = `
<View
  accessibilityState={
    Object {
      "disabled": true,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignContent": "center",
      "backgroundColor": "#67B6A8",
      "borderRadius": 8,
      "justifyContent": "center",
      "opacity": 1,
      "padding": 16,
      "width": "100%",
    }
  }
>
  <ActivityIndicator
    animating={true}
    color="white"
    hidesWhenStopped={true}
    size="small"
    style={
      Array [
        Object {},
      ]
    }
    testID="QBYQCQ-button-spinner"
  />
</View>
`;
